### PR TITLE
Determinism config and test

### PR DESCRIPTION
**Pull request by gobbleturk**
Add configurable determination option to maxtext and include a test for it

### CODE REVIEW

File: end_to_end/eval_assert.py [hunk 1]
```diff
@@ -44,6 +44,19 @@ def test_checkpointing(metrics_file, target):
...
+  run_2 = 'run_2_' + metrics_file
+
+  with open(run_1, 'r', encoding='utf8') as run_1_file,\
+    open(run_2, 'r', encoding='utf8') as run_2_file:
+    run_1_loss = json.loads(run_1_file.readlines()[-1])[target]
+    run_2_loss = json.loads(run_2_file.readlines()[-1])[target]
+    # Check that the two runs have the same loss
+    print(f"Run 1 loss:{run_1_loss}", flush=True)
+    print(f"Run 2 loss:{run_2_loss}", flush=True)
+    assert isclose(run_1_loss, run_2_loss, rel_tol=0.001)
```
Comment (Obliviour) [inline end_to_end/eval_assert.py:59]: Just to help my understanding, what is the reason for allowing a (small?) mismatch tolerance in the loss check here for determinism?

I would have initially thought that the values would be equivalent?

File: end_to_end/test_determinism_v4_16_two_slice.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+set -e
+
+RUN_NAME=${1}_$(date +%Y-%m-%d-%H)
+OUTPUT_PATH=${2}
+DATASET_PATH=${3}
+
+
+#Setup and Train
+#bash setup.sh
```
Comment (priyanka-ganesha) [inline end_to_end/test_determinism_v4_16_two_slice.sh:10]: Uncomment setup

**Review by priyanka-ganesha — COMMENTED** [review]
I believe there's also two references to enable_dropout in decode.py -- lines #84 and #100

File: end_to_end/test_determinism_v4_16_two_slice.sh
[the remark this answers is not in the document]
Comment (priyanka-ganesha) [reply end_to_end/test_determinism_v4_16_two_slice.sh:1]: Good question. The tests in the end_to_end folder are all run automatically nightly on XLML, which is our open-sourced testing framework. I'll ping you with some useful links privately!

File: end_to_end/test_determinism_v4_16_two_slice.sh (hunk 2 above)
Comment (gobbleturk) [reply end_to_end/test_determinism_v4_16_two_slice.sh]: Good catch! Fixed. I think its best to leave the two references to enable_dropout hardcoded to false in decode - for decoding there is no reason to use dropout (it is only a useful featuring for training)

**Review by priyanka-ganesha — APPROVED** [review]

File: end_to_end/test_determinism_v4_16_two_slice.sh
[the remark this answers is not in the document]
Comment (gobbleturk) [reply end_to_end/test_determinism_v4_16_two_slice.sh:1]: The tests in end_to_end don't get directly tested - they are referenced in XLML tests which do get tested nightly. I'll have you the review the associated XLML change (which lives outside of maxtext) for this test once this changed is pushed.

**Review by Obliviour — APPROVED** [review]

File: end_to_end/test_determinism_v4_16_two_slice.sh
[the remark this answers is not in the document]
Comment (Obliviour) [reply end_to_end/test_determinism_v4_16_two_slice.sh:1]: Thank you both for the great answers!
